[AMD][FA] Improve warp distribution for decode attention dot

zhanglx13 opened the pull request:
This PR improves the logic regarding warp distribution for FA kernels
1. Always choose warpsPerCTA=[numWarps, 1] for the 1st dot
2. For the 2nd dot, distribute warps along dim0 first, then dim1

This helps register pressure for FA kernel with a large output head size.

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp:
```diff
@@ -44,6 +44,50 @@ int getWmmaVersion(StringRef archGen) {
+// Check if the result of current tl.dot is used as the operand(0)
+// of another tl.dot
+bool isChainDotHead(tt::DotOp &dotOp) {
+  auto filter = [&dotOp](Operation *op) {
```
Typically we want to have self-documenting function/variable name for readability. It's not a big concern here given it's a oneliner, but `isInSameRegion` would be better than a general `filter` name here.

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp:
```diff
@@ -44,6 +44,50 @@ int getWmmaVersion(StringRef archGen) {
+  getForwardSlice(dotOp.getResult(), &fwdSlices, fwdOpt);
+  for (Operation *op : fwdSlices) {
+    // ensure output of the first dot is the operand 0 of the second dot
+    if (isa<tt::DotOp>(op) && (op != dotOp)) {
+      auto dOp = dyn_cast<tt::DotOp>(op);
```
`op` should never be `dotOp` here given we don't set `fwdOpt.inclusive` to `true`? To be certain you can put an `assert` in the above. Also typically in MLIR we do like

```c++
if (auto userDotOp = dyn_cast<tt::DotOpInterface>()) { ... }
```
Then in the middle you don't need to `cast` again.

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp:
```diff
@@ -44,6 +44,50 @@ int getWmmaVersion(StringRef archGen) {
+    // ensure output of the first dot is the operand 0 of the second dot
+    if (isa<tt::DotOp>(op) && (op != dotOp)) {
+      auto dOp = dyn_cast<tt::DotOp>(op);
+      auto op0 = dOp.getOperand(0).getDefiningOp();
```
Prefer to use friendly accessors like `.getA()`.

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp:
```diff
@@ -44,6 +44,50 @@ int getWmmaVersion(StringRef archGen) {
+    if (isa<tt::DotOp>(op) && (op != dotOp)) {
+      auto dOp = dyn_cast<tt::DotOp>(op);
+      auto op0 = dOp.getOperand(0).getDefiningOp();
+      if (op0 && std::find(fwdSlices.begin(), fwdSlices.end(), op0) !=
+                     fwdSlices.end()) {
```
You can use `fwdSlices.contains` here?

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp:
```diff
@@ -44,6 +44,50 @@ int getWmmaVersion(StringRef archGen) {
 
+// Check if the result of current tl.dot is used as the operand(0)
+// of another tl.dot
+bool isChainDotHead(tt::DotOp &dotOp) {
```
In these two functions, use `DotOpInterface` instead of hardcoded `tt::DotOp` so it works for `DotScaledOp` too. Also typically for friendly named `Op`s (that is, `tt::DotOp`, not `Operation *`), we don't pass as a reference; we directly pass as a value because `Op`s are just a wrapper of `Operation *`.

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp:
```diff
@@ -44,6 +44,50 @@ int getWmmaVersion(StringRef archGen) {
+
+// Check if the operand(0) of current tl.dot is the result of
+// another tl.dot
+bool isChainDotTail(tt::DotOp &dotOp) {
```
Similarly for this function.

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp:
```diff
@@ -52,20 +96,21 @@ warpsPerTile(Operation *dotOp, ArrayRef<int64_t> shape, int numWarps,
-  auto slices = getSlice(dotOp, bwdOpt, fwdOpt);
-  for (Operation *op : slices) {
-    if (isa<mlir::triton::DotOpInterface>(op) && (op != dotOp))
+  // For FA-like pattern, i.e. result of 1st tl.dot is used as the
+  // operand(0) of the 2nd dot.
+  // We use {numWaprs, 1} for both tl.dots
```
These three lines should be the same sentence.

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp:
```diff
@@ -52,20 +96,21 @@ warpsPerTile(Operation *dotOp, ArrayRef<int64_t> shape, int numWarps,
+  auto ttDotOp = dyn_cast<tt::DotOp>(dotOp);
+  if (isChainDotHead(ttDotOp) || isChainDotTail(ttDotOp)) {
+    if ((shape[0] == shapePerWarp.first) && isChainDotTail(ttDotOp))
+      return {1, (unsigned)numWarps};
```
We need to add a lit test for this.

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp:
```diff
@@ -52,20 +96,21 @@ warpsPerTile(Operation *dotOp, ArrayRef<int64_t> shape, int numWarps,
+  // {1, numWarps} for the 2nd tl.dot to save registers
+  auto ttDotOp = dyn_cast<tt::DotOp>(dotOp);
+  if (isChainDotHead(ttDotOp) || isChainDotTail(ttDotOp)) {
+    if ((shape[0] == shapePerWarp.first) && isChainDotTail(ttDotOp))
```
You can use a local variable to save the result for `isChainDotTail` to avoid compute it again. (The C++ compiler should do it but I'm not 100% sure.) Also this is a specific case. I suspect we want to swap here as long as it's more beneficial to distribute along second dot's N dim? That is, `ceildiv(shape[0], shapePerWarp.first) < ceildiv(shape[1], shapePerWarp.second)`?

zhanglx13 commented:
@antiagainst I addressed most of your comment except 1) the DotOpInterface for isChainDotHead function, since I need the `getC()` API from DotOpInterface. 2) lit tests, which I will add later.

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp:
```diff
@@ -43,28 +43,88 @@ int getWmmaVersion(StringRef archGen) {
+    SmallVector<unsigned, 3> ret = {1, 1};
+    ret[0] = static_cast<unsigned>(
+        std::min(static_cast<int64_t>(numWarps),
+                 (shape[0] + shapePerWarp.first - 1) / shapePerWarp.first));
```
Nit: this is `llvm::divideCeil`.

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp:
```diff
@@ -43,28 +43,88 @@ int getWmmaVersion(StringRef archGen) {
+
+  // Case 2: For FA-like pattern, i.e. result of 1st tl.dot is used as the opA
+  // of the 2nd dot, we will set warpsPerCTA differently for 1st and 2nd dot
+  auto ttDotOp = dyn_cast<tt::DotOpInterface>(dotOp);
```
Nit: can directly use `cast` here given the input is expected to be a dot.
